### ISSUE TITLE
Filesystem synchronisation to ensure durability

### ISSUE THREAD

**Issue by jamespharaoh** [zbackup/zbackup#136]
A quick search of the source code brings up nothing for `sync` or `fsync`. It seems to be that, in order to guarantee a backup has completed successfully then the following should take place:
- All files are written to a temporary location initially. I believe this is the case, not sure if it applies to every file though.
- An `fsync` is made on every one of these files.
- The files are moved into their new location, with the "backup" being the last of these.

I _believe_ that POSIX guarantees that a file rename (move) will be atomic and in-order, but I'd need to check that as well. There's probably some extra work needed to ensure things work with network filesystems, especially if there are multiple backing filesystems in a repository (which is what I am planning to do), but these are secondary I think.

I think the `fsync` is the main thing that is lacking. Apologies if my simple search has missed something here.

**Comment by mknjc** [comment]
Yes we should add fsyncs after writing the files, I'll look into it.
